### ISSUE TITLE
Support for DND 5e additive darkvision

### ISSUE THREAD

**Issue by Merudo** [RPTools/maptool#1895]
**Is your feature request related to a problem? Please describe.**
In 5e, Darkvision is additive: it turns Darkness into dim light, and dim light into bright light.

This type of vision cannot be currently implemented in MapTool.

**Describe the solution you'd like**
A extra parameter to sight, `add#rrggbb`, which would add the `#rrggbb` code to every light source seen by the token.

**Comment by JamzTheMan** [comment]
Are you going to add Pathfinder darkvision in as well? 😸 

It's only black/white (greyscale) vision and would look really cool for sight to be reduce to greyscale! Of course, it's only within range (normally 60/90/120ft) and any other light source in the area would overwrite/overpower/replace this within the lights range...

**Comment by JamzTheMan** [comment]
Oh and Pathfinder 2E lowlight (and other DnD versions) bump dim to bright as well.
Also, Darkness spells reduce bright to dim and dim to dark. But we can't do "that" right now (dim to dark, or dark to dim)

**Comment by melek** [comment]
I use dark red personal lights for darkvision (which washes the color out pretty well), and a bluish gray value for dim light; Something that has always irked me a bit in MT is that lights combine into darker colors, so my 'dim light in darkvision' looks darker than either dim light or darkvision. 

They _should_ add together to make brighter areas, which would resolve some of the weirdness.

MT Light (Subtractive):
![image](https://user-images.githubusercontent.com/220149/82581356-53d8fe80-9b45-11ea-928d-9f6493662171.png)

Actual Light (Additive):
![](http://cdn.differencebetween.net/wp-content/uploads/2018/06/Difference-Between-Additive-Colors-and-Subtractive-Colors-300x300.png)

**Comment by JamzTheMan** [comment]
@Phergus ya, why doesn't MapTool do it this way? Although, aren't we using RGB for lights and isn't that "additive" vs "subtractive" and should end up white?

I remember googling similar things doing the "color picker" for VBl and holy crap did I go down a rabbit hole and basically found out colors are hard lol

**Comment by Phergus** [comment]
My memories on this are fuzzy at best and I haven't looked at the code to see how light colors are being combined.  Given what we see in MT it wouldn't surprise me if we are doing a simple (R1+R2)/2 and so on which of course won't preserve luminance levels.

**Comment by melek** [comment]
Whatever is happening in MT right now is pretty weird.
Here are some values of the pure colored lights, 'Night' vision mode over pure white background drawing (which appears gray because of the slight night coloring).
```
// MT Definitions
Green: circle 20#00ff00
Red: circle 20#ff0000
Blue: circle 20#0000ff

// White base color under darkness, in RGB / HSL
White BG: RGB 210, 210, 210 / HSL 0° 0% 82% 

Red:   255, 155, 155 /   0° 39% 100%  // Pure colors, brighter than white base
Green:  55, 255, 155 / 120° 39% 100%
Blue:  155, 155, 255 / 240° 39% 100%

R+G:   194, 155,  94 /  37° 52%  76%   // Combined colors, dimmer than white base
R+B:   194,  94, 155 / 323° 52%  76%   // Why do these skew in favor of red/blue?
G+B:    94, 155, 194 / 203° 52%  76%

R+G+B: 157,  94, 118 / 337° 40%  62%  // Center. Muddy Red; Should be GRAY or WHITE
```

**Comment by JamzTheMan** [comment]
Took a quick peek, we're doing a graphics2d.fill()

So a quick google search led me to adding the setXORMode line above the render block here:

```    newG.setXORMode(Color.black);```
```
    for (Entry<Paint, List<Area>> entry : renderedLightMap.entrySet()) {
      newG.setPaint(entry.getKey());
      for (Area area : entry.getValue()) {
        newG.fill(area);
      }
    }
    timer.stop("lights-5");
    newG.dispose();
  }
```

And I got the below results... Promising.  I don't have time to chase it down further... We'll have to probably render the lights like this elsewhere in the code before adding the background image or something, but looks like we can at least set the coloring to additive a bit better...

![image](https://user-images.githubusercontent.com/8051654/82617252-1bf9a780-9b95-11ea-8bb3-0bae4dbf2da7.png)

vs the original

![image](https://user-images.githubusercontent.com/8051654/82617258-21ef8880-9b95-11ea-9df5-95ecbf9f215d.png)

**Comment by cwisniew** [comment]
You might want to try two overlapping lights of same colour before you go too far down the XOR path :)

**Comment by bubblobill** [comment]
Since you are fiddling with lights, I would suggest accommodating gradients for those that enjoy realism. Not saying you should implement gradients, just that if implemented they would work with any changes.
```
lamp: circle 20#ffffdd>#000000       linear
lamp: circle 20#ffffdd>>#000000    inverse square
lamp: circle 20#ffffdd>>>#000000  inverse cube
```

**Comment by JamzTheMan** [comment]
> Since you are fiddling with lights, I would suggest accommodating gradients for those that enjoy realism. Not saying you should implement gradients, just that if implemented they would work with any changes.
> 
> ```
> lamp: circle 20#ffffdd>#000000    linear
> lamp: circle 20#ffffdd>>#000000   inverse square
> lamp: circle 20#ffffdd>>>#000000  inverse cube
> ```

Think there is another issue for this but i like your notation...

**Comment by melek** [comment]
Yeah I made that issue for the same reasons @bubblobill suggested; figured it might be possible to accomodate if someone was messing with lights anyway ;)

I didn't want to dogpile this issue too badly but I can't say I wasn't tempted.. #1908

**Comment by Merudo** [comment]
Another topic that might be relevant: [Strange lighting when coloured light sources overlap](https://github.com/RPTools/maptool/issues/1550)

**Comment by rkyeun** [comment]
When light_source_1 of color <0.7 0.7 0.7> overlaps light_source_2 of color <0.5 0.0 0.0> the resulting color should be <1.0-((1.0-.07)*(1.0-0.5)) 1.0-((1.0-.07)*(1.0-0.0)) 1.0-((1.0-.07)*(1.0-0.0))>
When light_source_1 of color <0.7 0.7 0.7> overlaps light_source_2 of color <0.5 0.0 0.0> and a map art pixel with color <0.2 1.0 0.2> the resulting color should be <1.0-((1.0-.07)*(1.0-0.5)*(1-0.2)) 1.0-((1.0-.07)*(1.0-0.0)*(1.0-1.0)) 1.0-((1.0-.07)*(1.0-0.0)*(1-0.2))>
In general, inverse colors, multiply all overlaps together, inverse again. This makes the lights add.

**Comment by adyoungblood** [comment]
I've got an example working, implementing a Composite class as Craig mentioned. It additively blends colors as expected, and in conjunction with the optional #rrggbbhex option for Sights resolves the original issue as well. I'll touch up the code and get a pull request going as soon sometime this weekend.
![image](https://user-images.githubusercontent.com/16092423/149268850-9c13dac2-cc84-4c10-ae59-a79eeea2cf2a.png)

**Comment by rkyeun** [comment]
Neat! I had completely forgotten about this.

________________________________

**Comment by Phergus** [comment]
Testing.  Colored lights are now additive but is perhaps why darkness doesn't render at all.

Player View - Three low-saturation colored lights
![image](https://user-images.githubusercontent.com/34379254/178165034-2c1b898e-4907-4361-956a-82597da07430.png)

Player View - Three low-saturation colored lights.  With Darkness in center.
![image](https://user-images.githubusercontent.com/34379254/178165054-32c1b20d-a17a-43c6-bf48-dfe7f7995a25.png)

GM View
![image](https://user-images.githubusercontent.com/34379254/178165073-8306c98e-ac5d-4ca7-a72d-1d7bf3c5d5da.png)

See also #1550, #2094, #3346

**Comment by Phergus** [comment]
This issue was originally about making Darkvision enhancing existing lighting.  Making Darkness -> Dim and Dim -> Bright.  With current dev code on a GM client, light sources are brighter when no tokens are selected and get dimmer whenever any token (with or without Darkvision) is selected.  On a player client the same thing occurs.

**Comment by Phergus** [comment]
Player view token with Darkvision:
![image](https://user-images.githubusercontent.com/34379254/178555243-5c233a81-53c0-4616-9469-89680560259a.png)

Player view token with Normal vision:
![image](https://user-images.githubusercontent.com/34379254/178555340-ca459752-230f-4980-b6ca-0f7944366dea.png)

GM View:
![image](https://user-images.githubusercontent.com/34379254/178555415-1905424c-21bc-46fe-8934-3d1e66e6401c.png)

**Comment by bubblobill** [comment]
@kwvanderlinde Is this sorted now?

**Comment by kwvanderlinde** [comment]
Nope, although this FR was written in a time where dim/bright were primarily expressed by colours rather than lumens. So I don't think the OP really applies anymore, though the FR as a whole is still useful to keep around.

What we need for real darkvision is some kind of lumens-boosting capability. E.g., within 60' of a token with darkvision, all dim +50 lumens lights should become bright +100 lumens lights.
